packages/core/src/api/blockManipulation/commands/replaceBlocks/replaceBlocks.ts

```diff
@@ -70,26 +71,61 @@ export function removeAndInsertBlocks<
     }
 
     const oldDocSize = tr.doc.nodeSize;
-    // Checks if the block is the only child of its parent. In this case, we
-    // need to delete the parent `blockGroup` node instead of just the
-    // `blockContainer`.
+
     const $pos = tr.doc.resolve(pos - removedSize);
     if (
+      node.type.name === "column" &&
+      node.attrs.id !== $pos.nodeAfter?.attrs.id
+    ) {
+      // This is a hacky work around to handle removing all columns in a
+      // columnList. This is what happens when removing the last 2 columns:
+      //
+      // 1. The second-to-last `column` is removed.
+      // 2. `fixColumnList` runs, removing the `columnList` and inserting the
+      // contents of the last column in its place.
+      // 3. `removedSize` increases not just by the size of the second-to-last
+      // `column`, but also by the positions removed due to running
+      // `fixColumnList`. Some of these positions are after the contents of the
+      // last `column`, namely just after the `column` and `columnList`.
+      // 3. `tr.doc.descendants` traverses to the last `column`.
+      // 4. `removedSize` now includes positions that were removed after the
+      // last `column`. This causes `pos - removedSize` to point to an
+      // incorrect position, as it expects that the difference in document size
+      // accounted for by `removedSize` comes before the block being removed.
+      // 5. The deletion is offset by 3, because of those removed positions
+      // included in `removedSize` that occur after the last `column`.
+      //
+      // Hence why we have to shift the start of the deletion range back by 3.
+      // The offset for the end of the range is smaller as `node.nodeSize` is
+      // the size of the second `column`. Since it's been removed, we actually
+      // care about the size of its children - a difference of 2 positions.
+      tr.delete(pos - removedSize + 3, pos - removedSize + node.nodeSize + 1);
```

Comment:
I'm not certain how well this will all work. Does it really handle remove & inserting blocks properly? What about adding/removing within the column that may be "fix up"? Do we have test cases for this?

To avoid this:

Honestly, what I'd prefer is to do this in two passes:

1. where we do the full table scan, without trying to fix the columns along the way, capture the positions of any columns/columnLists (should just need columnLists right?)
2. do the column list cleanup afterwards for any columnLists found along the way

This avoids us needing to account for any changes in positioning based on the stuff we add/delete along the way. And fixes only after all of the changes have been applied (so we don't need to care where those changes are).